onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -39,9 +42,130 @@ static bool IsSmallInitializer(const onnxruntime::GraphViewer& graph, const Node
 }
 }  // namespace
 
+#if !defined(ORT_MINIMAL_BUILD) && !defined(ORT_EXTENDED_MINIMAL_BUILD)
+static InlinedHashSet<NodeIndex> GetShapeRelatedNodes(const onnxruntime::GraphViewer& viewer) {
+  // Conceptually, this function traverse from shape-consuming nodes
+  // to fallback all its upstream nodes to CPU. Consider a graph
+  //
+  //
+  // The traversal should stop when
+  //  1. hitting Shape, Size nodes, graph inputs, or graph initializers.
+  //  2. [TODO] hitting nodes with some large inputs or outputs. Before that,
+  //     we need shape inference to determine the size of the inputs and outputs.
+  //     Some graph transforms add nodes without shape information, so
+  //     checking shapes will make the algorithm more unstable now.
+  LOGS_DEFAULT(VERBOSE) << "Call GetShapeRelatedNodes to identify extra CPU nodes." << std::endl;
+
+  const static InlinedHashMap<std::string_view, InlinedHashMap<int64_t, std::vector<size_t>>> shape_related_inputs_in_nodes = {
```

Comment:
Curious to hear your thoughts on this - https://github.com/microsoft/onnxruntime/pull/19769/files#r1515337499